crates/pessimistic-proof-utils/src/smt.rs

```diff
@@ -0,0 +1,559 @@
+    #[serde_as(as = "[_; DEPTH]")]
+    empty_hash_at_height: [H::Digest; DEPTH],
+}
+
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SmtMerkleProof<H, const DEPTH: usize>
+where
+    H: Hasher,
+    H::Digest: Copy + Eq + Hash + Serialize + DeserializeOwned,
+{
+    #[serde_as(as = "[_; DEPTH]")]
+    siblings: [H::Digest; DEPTH],
+}
+
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SmtNonInclusionProof<H, const DEPTH: usize>
+where
+    H: Hasher,
+    H::Digest: Copy + Eq + Hash + Serialize + DeserializeOwned,
+{
+    #[serde_as(as = "Vec<_>")]
+    siblings: Vec<H::Digest>,
+}
+
+impl<H, const DEPTH: usize> Default for Smt<H, DEPTH>
+where
+    H: Hasher,
+    H::Digest: Copy + Eq + Hash + Serialize + DeserializeOwned + Default,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<H, const DEPTH: usize> Smt<H, DEPTH>
+where
+    H: Hasher,
+    H::Digest: Copy + Eq + Hash + Serialize + DeserializeOwned,
+{
+    pub fn new() -> Self
+    where
+        H::Digest: Default,
+    {
+        let empty_hash_at_height = empty_hash_at_height::<H, DEPTH>();
+        let root = H::merge(
+            &empty_hash_at_height[DEPTH - 1],
+            &empty_hash_at_height[DEPTH - 1],
+        );
+        let tree = HashMap::new();
+        Smt {
+            root,
+            tree,
+            empty_hash_at_height,
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.root
+            == H::merge(
+                &self.empty_hash_at_height[DEPTH - 1],
+                &self.empty_hash_at_height[DEPTH - 1],
+            )
+    }
+
+    pub fn get<K>(&self, key: K) -> Option<H::Digest>
+    where
+        K: ToBits<DEPTH>,
+    {
+        let mut hash = self.root;
+        for b in key.to_bits() {
+            hash = if b {
+                self.tree.get(&hash)?.right
+            } else {
+                self.tree.get(&hash)?.left
+            };
+        }
+
+        Some(hash)
+    }
+
+    fn insert_helper(
+        &mut self,
+        hash: H::Digest,
+        depth: usize,
+        bits: &[bool; DEPTH],
+        value: H::Digest,
+    ) -> Result<H::Digest, SmtError> {
+        if depth == DEPTH {
+            return if hash != self.empty_hash_at_height[0] {
+                Err(SmtError::KeyAlreadyPresent)
+            } else {
+                Ok(value)
+            };
+        }
+        let node = self.tree.get(&hash);
+        let mut node = node.copied().unwrap_or(Node {
+            left: self.empty_hash_at_height[DEPTH - depth - 1],
+            right: self.empty_hash_at_height[DEPTH - depth - 1],
+        });
+        let child_hash = if bits[depth] {
+            self.insert_helper(node.right, depth + 1, bits, value)
+        } else {
+            self.insert_helper(node.left, depth + 1, bits, value)
+        }?;
+        if bits[depth] {
+            node.right = child_hash;
+        } else {
+            node.left = child_hash;
+        }
```

Comment:
Could something like that works?

```suggestion
        let node_place = if bits[depth] { &mut node.right } else { &mut node.left }; 
        *node_place = self.insert_helper(*node_place, depth + 1, bits, value)?;
```

Reply:
I think so, very nice!